treeherder/perf/alerts.py

```diff
@@ -334,52 +337,96 @@ def _glance_over_retrigger_range(self, data_points_to_retrigger: List[dict]):
         :param since: datetime since the lookup will occur.
-        :param data_points_lookup_interval: time range before/after data point in which search neighboring data points occurs.
         '''
         self.since = since
-        self.picker = AlertsPicker(
-            max_alerts=5,
-            max_improvements=2,
-            platforms_of_interest=('windows10', 'windows7', 'linux', 'osx', 'android'))
-        self.find_alert_retriggerables = IdentifyAlertRetriggerables(
-            max_data_points=5,
-            time_interval=data_points_lookup_interval)
-
-    def __call__(self, frameworks: List[str], repositories: List[str]) -> dict:
+        self.alerts_picker = alerts_picker
+        self.fetch_backfill_context = backfill_context_fetcher
+        self.log = logger or logging.getLogger(self.__class__.__name__)
+
+    def provide_updated_reports(self, frameworks: List[str], repositories: List[str]) -> List[BackfillReport]:
         summaries_to_retrigger = self._fetch_by(
-            self._summaries(self.since),
+            self._summaries_requiring_reports(self.since),
             frameworks,
             repositories
         )
-        return self._identify_retriggerables(summaries_to_retrigger)
+        return self.compile_reports_for(summaries_to_retrigger)
+
+    def compile_reports_for(self, summaries_to_retrigger: QuerySet) -> List[BackfillReport]:
+        reports = []
+
+        for summary in summaries_to_retrigger:
+            important_alerts = self._pick_important_alerts(summary)
+            if len(important_alerts) == 0 and self._doesnt_have_report(summary):
+                continue  # won't create blank reports
+            # but will update if case
+
+            try:
+                alert_context_map = self._associate_retrigger_context(important_alerts)
+            except MissingRecords as ex:
+                self.log.warning(f"Failed to compute report for alert summary {summary}. {ex}")
+                continue
+
+            backfill_report, created = BackfillReport.objects.get_or_create(summary_id=summary.id)
+            if created or backfill_report.is_outdated:
+                self._provide_records(backfill_report, alert_context_map)
+            reports.append(backfill_report)
+
+        return reports
+
+    def _pick_important_alerts(self, from_summary: PerformanceAlertSummary) -> List[PerformanceAlert]:
+        return self.alerts_picker.extract_important_alerts(from_summary.alerts.all())
+
+    def _provide_records(self, backfill_report: BackfillReport,
+                         alert_context_map: List[Tuple]):
+        # any existing records are outdated; remove them
+        backfill_report.expel_records()
+
+        for alert, retrigger_context in alert_context_map:
+            BackfillRecord.objects.create(alert=alert,
+                                          report=backfill_report,
+                                          context=json.dumps(retrigger_context))
```

Comment:
> It wouldn't be more appropriate to use a custom serializer instead of json.dumps? Just asking.

I'm not sure this is what Django serializers where created for.

Reply:
I think this is what serializers are useful for.